packages/storage/src/reference.ts

```diff
@@ -201,6 +203,89 @@ export class Reference {
     });
   }
 
+  /**
+   * List all items (files) and prefixes (folders) under this storage reference.
+   *
+   * It is a helper method for calling list() recursively until there is no
+   * more results. The default pagination size is 200.
+   *
+   * Note: the results may not be a consistent snapshot if objects are changed
+   * between paginating requests.
+   *
+   * Warning: If there are
+   *
+   * @return A promise that resolves with all the items and prefixes under
+   *      the current storage reference. `prefixes` contains references to
+   *      sub-directories and `items` contains references to objects in this
+   *      folder. `nextPageToken` is never returned.
+   */
+  listAll(): Promise<ListResult> {
+    args.validate('listAll', [], arguments);
+    const accumulator = {
+      prefixes: [],
+      items: []
+    };
+    return this.listAllHelper(accumulator, null).then(() => accumulator);
+  }
+
+  private async listAllHelper(
+    accumulator: ListResult,
+    pageToken?: string
+  ): Promise<void> {
+    let opt: ListOptions = {
+      maxResults: 1,
+      pageToken
+    };
+    const nextPage = await this.list(opt);
+    Array.prototype.push.apply(accumulator.prefixes, nextPage.prefixes);
```

Comment:
The new and modern syntax for this is `accumulator.prefixes(...nextPage.prefixes)`, which runs in O(n).

You could also use `concat` and make this potentially faster.

Reply:
Done.
I thought `concat` does not modify the existing list, so it is O(n2) for many pages.